File: apps/ios/Janark/Core/Time.swift
import Foundation

private let enIN = Locale(identifier: "en_IN")

private let shortDateFormatter: DateFormatter = {
    let f = DateFormatter()
    f.locale = enIN
    f.dateStyle = .medium
    f.timeStyle = .none
    return f
}()

func parseInstant(_ iso: String) -> Date? {
    let isoFormatter = ISO8601DateFormatter()
    isoFormatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
    if let d = isoFormatter.date(from: iso) { return d }
    isoFormatter.formatOptions = [.withInternetDateTime]
    if let d = isoFormatter.date(from: iso) { return d }
    let patched = iso.replacingOccurrences(of: " ", with: "T") + "Z"
    return isoFormatter.date(from: patched)
}

func relativeTime(_ iso: String?) -> String? {
    guard let iso, !iso.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else { return nil }
    guard let t = parseInstant(iso) else {
        return String(iso.prefix(10))
    }
    let sec = Int(Date().timeIntervalSince(t))
    if sec < 60 { return "Just now" }
    let min = Int((Double(sec) / 60.0).rounded())
    if min < 60 { return "\(min)m ago" }
    let hr = Int((Double(min) / 60.0).rounded())
    if hr < 48 { return "\(hr)h ago" }
    let day = Int((Double(hr) / 24.0).rounded())
    if day < 14 { return "\(day)d ago" }
    return shortDateFormatter.string(from: t)
}

func formatDate(_ iso: String?) -> String {
    guard let iso, !iso.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else { return "" }
    guard let t = parseInstant(iso) else { return String(iso.prefix(10)) }
    return shortDateFormatter.string(from: t)
}

func placeLabel(
    city: String? = nil,
    district: String? = nil,
    state: String? = nil,
    country: String? = nil,
    extra: String? = nil
) -> String? {
    let parts = [extra, city, district, state]
        .compactMap { $0?.trimmingCharacters(in: .whitespacesAndNewlines) }
        .filter { !$0.isEmpty }
    let label = parts.joined(separator: ", ")
    if !label.isEmpty { return label }
    let c = country?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
    return c.isEmpty ? nil : c
}

func indianCount(_ n: Int) -> String {
    let formatter = NumberFormatter()
    formatter.locale = enIN
    formatter.numberStyle = .decimal
    return formatter.string(from: NSNumber(value: n)) ?? "\(n)"
}

File: apps/android/app/src/main/java/org/janark/app/core/Time.kt
package org.janark.app.core

import java.time.Instant
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import java.time.format.FormatStyle
import java.util.Locale

private val enIn = Locale("en", "IN")
private val shortDate = DateTimeFormatter.ofLocalizedDate(FormatStyle.MEDIUM).withLocale(enIn)

fun relativeTime(iso: String?): String? {
    if (iso.isNullOrBlank()) return null
    val t = parseInstant(iso) ?: return iso.take(10)
    val sec = ((System.currentTimeMillis() - t.toEpochMilli()) / 1000.0).toInt()
    if (sec < 60) return "Just now"
    val min = Math.round(sec / 60.0).toInt()
    if (min < 60) return "${min}m ago"
    val hr = Math.round(min / 60.0).toInt()
    if (hr < 48) return "${hr}h ago"
    val day = Math.round(hr / 24.0).toInt()
    if (day < 14) return "${day}d ago"
    return shortDate.format(t.atZone(ZoneId.systemDefault()).toLocalDate())
}

fun formatDate(iso: String?): String {
    if (iso.isNullOrBlank()) return ""
    val t = parseInstant(iso) ?: return iso.take(10)
    return shortDate.format(t.atZone(ZoneId.systemDefault()).toLocalDate())
}

fun parseInstant(iso: String): Instant? = runCatching {
    Instant.parse(iso)
}.getOrElse {
    runCatching { Instant.parse(iso.replace(" ", "T") + "Z") }.getOrNull()
}

fun placeLabel(
    city: String? = null,
    district: String? = null,
    state: String? = null,
    country: String? = null,
    extra: String? = null,
): String? {
    val label = listOfNotNull(extra, city, district, state).filter { it.isNotBlank() }.joinToString(", ")
    return label.ifBlank { country?.takeIf { it.isNotBlank() } }
}

fun indianCount(n: Int): String = "%,d".format(enIn, n)
